### PR TITLE
feat(env): consolidate onto cli-engine's shared Environments, wire feature-flag stages (DEVEX-929)

### PR DESCRIPTION
**Pull request by jpage-godaddy**
## Summary
- Replaces gddy's hand-rolled `rust/src/environments/mod.rs` resolution logic with cli-engine 0.4.8's shared `Environments`/`EnvironmentDef` type (bumped from 0.4.7), while keeping the module's own path/API surface stable so downstream call sites needed no import changes.
- Builds each `PkceAuthProvider` per call from gddy's own resolved (and gddy-specific-derived) endpoints, wired via `.with_environments` for the `<ENV>_OAUTH_*` override layer. The provider is always named the fixed `"godaddy"` (not the target env), not gddy's former per-env-named providers — this changes cli-engine's credential storage key from `(app_id, env, env)` to `(app_id, "godaddy", env)`. **Existing stored OAuth tokens are orphaned; users need to `gddy auth login` again after this lands.** PATs are unaffected.
- Fixes a live bug found along the way: `hosting`/`webhook`/`application`/`actions_catalog` were completely hidden in every build (`gddy hosting` → "unknown command") because nothing ever set `min_stage` below cli-engine's `Stage::Ga` default, despite being advertised in the CLI's own root `--help` text. `main.rs` now sets this explicitly, and confirmed end-to-end that an environment's resolved `min_stage`/`feature_overrides` (via a local `environments.toml` entry or `<ENV>_MIN_STAGE`/`<ENV>_FEATURE_<KEY>` env vars) can reveal them per-environment — this was already automatic once `CliConfig::with_environments` was wired, no new engine-side plumbing needed. This only takes effect for the environment active when the process *starts* (persist via `gddy env set <name>`, not a same-invocation `--env`).
- Preserves gddy-specific behavior cli-engine's `Environments` has no notion of: `domains_api_url`/`account_url` defaulting, `auth_url`/`token_url` derivation from `api_url`, URL validation (a blank/malformed override — including on the OAuth URLs — is rejected rather than silently used), and defining a brand-new environment purely via `<PREFIX>_API_URL` with no compiled/file entry (gddy's own pre-existing feature; cli-engine's own env-var layer can't discover a name on its own). The env-var-only-environment mechanism also validates a corrupted/stale `.gdenv` default (falls back to `prod` rather than becoming an unresolvable startup default) and correctly disambiguates the `--env` value from raw argv (last occurrence wins when both a global and command-local `--env` are present, a following flag is never mistaken for a value, and it scans safely even if an argument isn't valid UTF-8).

Related engine-side gaps found during this work were spun out into DEVEX-947 and DEVEX-948 rather than blocking this PR.

Jira: DEVEX-929

## Test plan
- [x] `cargo build`, `cargo test` (353 passing), `cargo clippy --all-targets -- -D warnings`, `cargo fmt --check` — all clean
- [x] Manually verified `env list/get/set/info`, `--env` validation and the env-var-only-environment fallback (including multiple `--env` occurrences and a blank/malformed override), against the real built binary
- [x] Manually verified a local `environments.toml`'s nested `[environments.<name>]` shape still parses with zero edits
- [x] Manually verified `min_stage`/`feature_overrides` in `environments.toml` reveal gated modules end-to-end, using (and restoring) the real local config file
- [x] Manually verified (`RUST_LOG=gddy=debug`) that the real `dev`/`test` environments — which only set `client_id` — resolve to the correctly-derived `auth_url`/`token_url` for the actual OAuth flow, not just for logging
- [x] Regression tests added throughout the review loop for each of the above (argv parsing edge cases, `.gdenv` fallback, OAuth URL validation, stage-gated module visibility)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: rust/src/environments/mod.rs [hunk 1]
```diff
@@ -179,171 +160,171 @@ fn clean_url(raw: &str) -> Option<String> {
...
+    let auth_url = if oauth.auth_url.is_empty() {
+        derive_auth_url(&api_url)
+    } else {
+        oauth.auth_url
+    };
+    let token_url = if oauth.token_url.is_empty() {
+        derive_token_url(&api_url)
+    } else {
+        oauth.token_url
+    };
```
[the remark this answers is not in the document]
Comment (jpage-godaddy) [reply rust/src/environments/mod.rs]: Fixed in 2a448b66df2dce5ad61d34aa5dc71c3434a33b81: auth_url/token_url now go through the same `clean_url` validation as api_url/domains_api_url/account_url, so a blank-but-non-empty or schemeless override falls back to the derived endpoint instead of being used as-is. Added a regression test (`adapt_falls_back_to_derived_oauth_urls_when_override_is_blank_or_malformed`).

File: rust/src/environments/mod.rs [hunk 2]
```diff
@@ -179,171 +160,168 @@ fn clean_url(raw: &str) -> Option<String> {
...
+/// cli-engine's built-in global `--env` flag (and any command-local `--env`
+/// arg sharing that argv text, e.g. `pat add --env <name>`) validates against
+/// the shared `Environments` *directly* — bypassing this module's own
+/// [`resolve`] and its env-var-only fallback entirely. So a name defined only
+/// via `<PREFIX>_API_URL` must be pre-registered into the singleton itself
+/// (see [`build_environments`]) before cli-engine ever sees it, which means
+/// finding that name before clap has parsed anything.
+fn requested_env_from_argv() -> Option<String> {
+    requested_env_from(std::env::args().skip(1))
 }
```
[the remark this answers is not in the document]
Comment (jpage-godaddy) [reply rust/src/environments/mod.rs:183]: Fixed in a464c1869c54a833856a269342eb15c867f52876: switched to `std::env::args_os()` with a lossy conversion, so a non-UTF-8 argument no longer panics during the environment singleton's initialization.

File: rust/src/environments/mod.rs [hunk 3]
```diff
@@ -179,171 +160,177 @@ fn clean_url(raw: &str) -> Option<String> {
...
+/// this module's own resolution helpers below.
+fn build_environments() -> Environments {
+    // `crate::env::read_gdenv_raw`, not `crate::env::get_env` — `get_env`
+    // validates via `is_known`, which resolves against this very instance and
+    // would deadlock re-entering `instance()`'s `OnceLock`
+    // mid-initialization.
+    let default = crate::env::read_gdenv_raw().unwrap_or_else(|| DEFAULT_ENV.to_owned());
+    let mut envs = Environments::new(default.clone())
+        .with_app_id(APP_ID)
+        .with_config_file(true);
```
[the remark this answers is not in the document]
Comment (jpage-godaddy) [reply rust/src/environments/mod.rs]: Fixed in b5dda3bc989803ae06245e55ac8c2fef2d3eab62: extracted the default-selection into resolve_default_environments, which now validates the .gdenv value against a local probe instance and falls back to DEFAULT_ENV if it can't resolve — matching env::get_env's own is_known guard, so every command reading ctx.middleware.env directly stays consistent with env get/info. Added two regression tests.

File: rust/src/env/mod.rs [hunk 4]
```diff
@@ -153,7 +159,11 @@ pub fn module() -> Module {
                     "Switches the active environment and persists the choice to \
                          ~/.gdenv so all subsequent commands use the new environment \
                          without needing `--env`.\n\
-                         The value must be a known environment name (prod or ote).",
+                         The value must be a known environment name (prod or ote).\n\
+                         If a local environments.toml entry for this environment sets \
+                         min_stage/feature_overrides to reveal pre-release commands, \
+                         persisting it here (not a same-invocation `--env`) is what \
+                         makes those commands visible on the next run.",
```
[the remark this answers is not in the document]
Comment (jpage-godaddy) [reply rust/src/env/mod.rs]: Fixed in b5dda3bc989803ae06245e55ac8c2fef2d3eab62: reworded to mention environments.toml entries and <PREFIX>_API_URL-defined environments, not just the two built-ins.

File: rust/src/environments/mod.rs [hunk 5]
```diff
@@ -179,171 +160,207 @@ fn clean_url(raw: &str) -> Option<String> {
...
+        if envs.list().contains(&candidate) {
+            continue;
         }
-        if let Some(cid) = &entry.client_id {
-            client_id = cid.clone();
+        let prefix = env_prefix(&candidate);
+        if std::env::var(format!("{prefix}_API_URL")).is_ok() {
+            envs =
+                envs.with_environment(candidate, EnvironmentDef::new().with_field("api_url", ""));
         }
```
[the remark this answers is not in the document]
Comment (jpage-godaddy) [reply rust/src/environments/mod.rs:319]: Fixed in d83e642890b937b04c134aa35023a57c5fc64965: gated the placeholder registration on clean_url validity of the env var's value instead of a bare .ok(), consistent with every other URL check in this module. Verified end-to-end: a blank FOO_API_URL now correctly fails with "unknown environment" instead of being accepted then failing later with a confusing error.

File: rust/src/environments/mod.rs [hunk 6]
```diff
@@ -179,171 +160,216 @@ fn clean_url(raw: &str) -> Option<String> {
...
+    while let Some(arg) = args.next() {
+        if let Some(value) = arg.strip_prefix("--env=") {
+            return Some(value.to_owned());
+        }
+        if arg == "--env" {
+            return args.next();
+        }
     }
+    None
 }
```
[the remark this answers is not in the document]
Comment (jpage-godaddy) [reply rust/src/environments/mod.rs:219]: Fixed in 79d698f1d89e89942416d4dd6ae8b8b7b808d9fc: requested_env_from now scans the entire argv and keeps the last non-empty --env/--env= value, matching cli-engine's own resolution (empirically confirmed the last occurrence wins for a shared arg id), and ignores an empty --env= instead of treating it as a literal empty-string candidate. Reproduced the exact scenario (a global --env plus a command-local --env with a different, env-var-only-defined value) against the built binary before and after the fix. Added regression tests for last-wins and empty-value handling.

File: rust/src/environments/mod.rs [hunk 7]
```diff
@@ -179,171 +160,232 @@ fn clean_url(raw: &str) -> Option<String> {
...
+fn requested_env_from(mut args: impl Iterator<Item = String>) -> Option<String> {
+    let mut result = None;
+    while let Some(arg) = args.next() {
+        let value = if let Some(v) = arg.strip_prefix("--env=") {
+            Some(v.to_owned())
+        } else if arg == "--env" {
+            args.next()
+        } else {
+            None
+        };
```
[the remark this answers is not in the document]
Comment (jpage-godaddy) [reply rust/src/environments/mod.rs:213]: Fixed in 9c10c2d1b63d4d5cbc88c5dd9b0c39ae5ad836ed: a space-separated value starting with - is no longer treated as --env's value, matching clap's own rejection of --env --dry-run (confirmed empirically: clap errors with "a value is required for '--env <ENV>' but none was supplied"). --env=-foo still works, matching clap's own disambiguation of the = form. Added regression tests for both.

File: rust/src/auth.rs [hunk 8]
```diff
@@ -8,47 +10,44 @@ use crate::environments::{self, ResolvedEnv};
...
 impl GoDaddyAuthProvider {
     pub fn new() -> Self {
-        Self
+        Self {
+            provider: PkceAuthProvider::new(
+                "godaddy",
+                "",
+                "",
+                "",
+                environments::DEFAULT_OAUTH_SCOPES,
```
[the remark this answers is not in the document]
Comment (jpage-godaddy) [reply rust/src/auth.rs]: Confirmed and fixed in da81c2614f9d18355cd24db7aec5322f0e3634ac -- this was real, not hypothetical: the real dev/test file entries only set client_id, so with empty base args the actual OAuth flow would resolve to an empty auth_url/token_url (confirmed via RUST_LOG=gddy=debug before the fix). Reverted to building the PkceAuthProvider per call from crate::environments::resolve's fully-derived values, but kept the provider name fixed as "godaddy" (not the env name) so the credential storage key stays the single stable identity from the earlier simplification -- verified end-to-end that the debug log now shows the correctly-derived auth_url/token_url for --env dev.

File: rust/src/env/mod.rs [hunk 9]
```diff
@@ -104,7 +110,7 @@ pub fn module() -> Module {
                 .no_auth(true),
             |_cred, _args| async move {
                 let current = active_env();
-                let mut resolved = environments::listable().map_err(map_err)?;
+                let mut resolved = environments::listable()?;
                 // If the active env is defined only via `<ENV>_API_URL` (so it's
                 // excluded from `listable`), still show it — otherwise the list
```
[the remark this answers is not in the document]
Comment (jpage-godaddy) [reply rust/src/env/mod.rs:116]: Fixed in da81c2614f9d18355cd24db7aec5322f0e3634ac: reworded to mention the built-in prod/ote plus any environments.toml entries.

File: rust/src/environments/mod.rs [hunk 10]
```diff
@@ -179,171 +160,238 @@ fn clean_url(raw: &str) -> Option<String> {
...
+/// initialization. Split out from [`build_environments`] for testability
+/// without touching the real `.gdenv` file or the process-wide singleton.
+fn resolve_default_environments(default_raw: &str) -> Environments {
+    let probe = register(Environments::new(default_raw.to_owned()), default_raw);
+    if probe.resolve(default_raw).is_ok() {
+        probe
+    } else {
+        register(Environments::new(DEFAULT_ENV), default_raw)
+    }
 }
```
[the remark this answers is not in the document]
Comment (jpage-godaddy) [reply rust/src/environments/mod.rs:265]: Fixed in 8c548c34b6bd149432b720a6bb72b9b2bcec9fde: build_environments() is now the only real caller of requested_env_from_argv(); resolve_default_environments and register take the scanned value as an explicit Option<String> parameter instead of calling it themselves, so they're deterministic for tests regardless of the test binary's own argv. Added a test proving the injected value is used without touching real argv.

**Review by qcai-godaddy — APPROVED** [review]
